### PR TITLE
Add json parsing for dps in profileset simulation

### PR DESCRIPTION
**Pull request by scamille**
Attempt at replacing the existing stdout parsing for the simulationGroup simulate function with json report parsing.

No attempt at parsing more than the dps for now.

### CODE REVIEW

File: bloodytools/simulation_objects/simulation_objects.py [hunk 1]
```diff
@@ -653,19 +655,16 @@ def simulate(self) -> bool:
...
-            self.logger.info(
-              "Somehow the random filename generator generated a name ('{}') that is already on use.".format(
-                self.filename
-              )
-            )
-            self.filename = str(uuid.uuid4()) + ".simc"
           # write arguments to file
           with open(self.filename, "w") as f:
             # write the equal values to file
+            f.write("json={}\n".format(json_filename))
```
Comment (Bloodmallet) [inline bloodytools/simulation_objects/simulation_objects.py]: Aren't we using `json2=` in SimulationCraft now?

File: bloodytools/bloodytools.py [hunk 2]
```diff
@@ -1222,7 +1222,7 @@ def main():
   if args.debug:
     settings.debug = args.debug
     logger.setLevel(logging.DEBUG)
-    ch.setLevel(logging.DEBUG)
+    console_handler.setLevel(logging.DEBUG)
```
Comment (Bloodmallet) [inline bloodytools/bloodytools.py:1225]: 👀 good catch

File: bloodytools/simulation_objects/simulation_objects.py [hunk 3]
```diff
@@ -653,19 +655,16 @@ def simulate(self) -> bool:
...
+            f.write("profileset_metric={}\n".format(",".join(["dps"])))
```
Comment (Bloodmallet) [inline bloodytools/simulation_objects/simulation_objects.py:668]: I have no clue
1) what a `profileset_metric` is
2) why you wrote that line as you did? support for extraction of more information, separated by comma?

File: bloodytools/simulation_objects/simulation_objects.py [hunk 4]
```diff
@@ -653,19 +655,16 @@ def simulate(self) -> bool:
             .format(self.filename)
           )
         else:
-          self.filename = str(uuid.uuid4()) + ".simc"
+          # temporary file names
+          self.uuid = str(uuid.uuid4())
+          self.filename = "{}.simc".format(self.uuid)
+          json_filename = "{}.json".format(self.uuid)
```
Comment (Bloodmallet) [inline bloodytools/simulation_objects/simulation_objects.py]: Please add `self.`

File: bloodytools/simulation_objects/simulation_objects.py [hunk 5]
```diff
@@ -1237,6 +1213,18 @@ def simulate_with_raidbots(self, apikey) -> bool:
 
     return simc_hash
 
+  def set_json_data(self, data: dict):
```
Comment (Bloodmallet) [inline bloodytools/simulation_objects/simulation_objects.py]: `def set_json_data(self, data: dict) -> None:`

Comment (Bloodmallet) [reply bloodytools/simulation_objects/simulation_objects.py]: I'm sure return type predictions are missing elsewhere in my code, too. But I think it's good to add it whenever something like this is created or edited.

**Review by Bloodmallet — CHANGES_REQUESTED** [review]
Please see the comments. Especially the `json` vs `json2` one

File: bloodytools/simulation_objects/simulation_objects.py (hunk 1 above)
Comment (scamille) [reply bloodytools/simulation_objects/simulation_objects.py]: I think there is no difference between the 2.
https://github.com/simulationcraft/simc/blob/bfa-dev/engine/sim/sc_sim.cpp#L3378
looks like the old json report died.

File: bloodytools/bloodytools.py (hunk 2 above)
Comment (scamille) [reply bloodytools/bloodytools.py:1225]: well it yelled at me when I ran it ;)

File: bloodytools/simulation_objects/simulation_objects.py (hunk 3 above)
Comment (scamille) [reply bloodytools/simulation_objects/simulation_objects.py:668]: Well SimC has this metric concept, which is mostly a reference to some collected data sample_data object.
dps, dpse, hps, hpse, dtps, tmi, etc.

Now with normal sims, you get all of them collected (although with different amount of detail, depending on the statistics_level).

With profileset because of the amount of data, only dps is collected by default.

The idea would be to add eg. dps, hps, tmi here. For now I reverted it to just 'dps'.
So basically this line has no effect for now.

File: bloodytools/simulation_objects/simulation_objects.py (hunk 4 above)
Comment (scamille) [reply bloodytools/simulation_objects/simulation_objects.py]: ok so in init I would add it as
self.json_filename : str = None
Correct?

Is there a reason for adding it to the class, as opposed to a local variable?

Comment (Bloodmallet) [reply bloodytools/simulation_objects/simulation_objects.py]: Sounds good. Yes the reason is, that I intend to keep simulation-related files. So input file, result-json, and the processed result-json will all be available. Therefore I'd like to keep that name around already.

File: bloodytools/simulation_objects/simulation_objects.py [hunk 6]
```diff
@@ -784,31 +785,17 @@ def simulate(self) -> bool:
...
-            if "Profilesets (median Damage per Second):" in line:
-              profileset_results = True
-              continue
 
-            # prevents false positive from Waiting -data
-            if "" == line and profileset_results:
-              profileset_results = False
-              continue
+          # parse results from generated json file
+          with open(self.json_filename) as json_file:
```
Comment (Bloodmallet) [inline bloodytools/simulation_objects/simulation_objects.py]: please add the mode, too. e.g.
```python
with open('my_file.txt', 'r') as f:
```

**Review by Bloodmallet — CHANGES_REQUESTED** [review]
I'll look into cloning your version locally tomorrow. One final thing: I like to be explicit with the open function

**Comment by scamille** [comment]
Do you have plans to support the non-profileset simulation function

```
      if len(self.profiles) == 1:
        # if only one profiles is in the group this profile is simulated normally
        try:
          self.profiles[0].simulate()
        except Exception as e:
          raise e
```

Or does the profileset setup not work in the special case of having only 1 profile?

I explicitly only worked on this part here, but might be worth consolidating things before adding further json parsing.

**Comment by Bloodmallet** [comment]
It would be great if the setup would still be able to handle single-profile-simulations, too.
